### PR TITLE
Refactor SimulationData normalization once again

### PR DESCRIPTION
**Pull request by momchil-flex**
`SimulationData` is now such that the normalization (still corresponding to the recorded `normalize_index`) is incorporated in the stored data. A new `renormalize` method is introduced to create a copy of the `SimulationData` with a different normalization. 

This has the following benefits:

- Avoids the confusion users may have of why the data in the hdf5 file is different from the data in python. There will always be a one-to-one match. 
- On the backend, we will use normalize_index = 0 by default. This is what the user downloads, and also can directly be used for post-run viz. Then the user can use the `renormalize` method if they want to remove the normalization (`renormalize(normalize_index=None)`) or pick a different source.
- Together with a change I made for `MonitorData.apply_symmetry` to only make a copy if any symmetries are actually present, now `sim_data["monitor_name"]` will no longer make a copy of the data (more memory efficient) (unless symmetries are present).

I also reverted the `apply_symmetry` method to not use `simulation` but only the necessary quantities.

### CODE REVIEW

File: tidy3d/components/data/sim_data.py [hunk 1]
```diff
@@ -145,7 +142,42 @@ def source_spectrum_fn(freqs):
...
+            new_spectrum_fn = self.source_spectrum(normalize_index)
+            old_spectrum_fn = self.source_spectrum(self.normalize_index)
+            return new_spectrum_fn(freqs) / old_spectrum_fn(freqs)
+
+        # Make a new monitor_data dictionary with renormalized data
+        monitor_data = {}
+        for key, val in self.monitor_data.items():
+            monitor_data[key] = val.normalize(source_spectrum_fn)
+
+        return self.copy(update=dict(normalize_index=normalize_index, monitor_data=monitor_data))
```
Comment (tylerflex) [inline tidy3d/components/data/sim_data.py:180]: See my other comment in the original hdf5 performance PR, but doesn't this just defeat the purpose as in this line there are two `SimulationData` instances in memory at the same time? Maybe I'm just confused on this point: does a `return self.copy()` call hold two copies in memory? or is smart enough to free the memory for the original one before allocating for the returned one?

Comment (momchil-flex) [reply tidy3d/components/data/sim_data.py:180]: So, I see your point, I would expect this to double the memory. However, the point of this PR is that in the vast majority of cases, people don't actually change the normalization: they just use the default `normalize_index = 0`. This is how the data will be provided from the server, so when they do `SimulationData.from_file`, no memory duplication will happen. I think the point is that while the less likely scenario of `sim_data.renormalize` does have memory doubling, this PR avoids memory doubling in the more likely scenario in which you have one very large monitor, and you do `sim_data["large_monitor"]`.

**Comment by tylerflex** [comment]
I see. We could also try adding normalize index=0 to the simulation itself
and just leaving it as a parameter that gets set once and doesn’t change

**Comment by momchil-flex** [comment]
Well, in what way is this better than what's done in this PR? I guess just that it is clear from the start what index is used? But it sacrifices the ability to renormalize easily (including remove normalization) if a user really wanted that.

**Comment by tylerflex** [comment]
It just means that the user needs to decide the normalization index before
submission. Then it’s fixed and we don’t need to copy or anything

**Comment by momchil-flex** [comment]
But why is this better? The copy to renormalize to index 0 only happens once on our servers, where we generally don't worry about running out of memory.

**Comment by momchil-flex** [comment]
Oh I guess you mean if a user *did* want a different index, they need to renormalize locally.

**Comment by momchil-flex** [comment]
Well, we could add the normalize_index as a Field in Simulation *and* keep the `renormalize` function? Essentially move `normalize_index` from `SimulationData` to `Simulation`?

**Comment by tylerflex** [comment]
I guess in my mind the user could just need to decide before submission
which source they’re normalizing by. And then there is no option to
renormalize. (Or maybe they could, but it might require a lot of memory).

**Comment by momchil-flex** [comment]
I moved `normalize_index` to `Simulation` but the user still has the option to `renormalize`. Also the `renormalize` function is how I will implement the selected `normalize_index` on the backend. How's this now?

File: tidy3d/components/data/monitor_data.py [hunk 2]
```diff
@@ -65,28 +68,24 @@ def grid_locations(self) -> Dict[str, str]:
...
+        expanded. If symmetry not present, returns the same data object.
 
         Returns
         -------
         :class:`AbstractFieldData`
-            A new data object with the symmetry expanded fields.
+            A data object with the symmetry expanded fields.
         """
 
+        if np.all(symmetry == 0):
```
Comment (tylerflex) [inline tidy3d/components/data/monitor_data.py:86]: This will always evaluate to `False` because `symmetry` is a tuple.  I think instead we want
```py
if all(s==0 for s in symmetry)
```

File: tidy3d/components/simulation.py [hunk 3]
```diff
@@ -180,6 +180,13 @@ class Simulation(Box):  # pylint:disable=too-many-public-methods
         "based on structure definition, resulting in much higher accuracy for a given grid size.",
     )
 
+    normalize_index: Optional[pydantic.NonNegativeInt] = pydantic.Field(
+        0,
+        title="Normalization index",
+        description="Index of the source in the list of sources whose spectrum will be used to "
```
Comment (tylerflex) [inline tidy3d/components/simulation.py:186]: "tuple" of sources, maybe

Comment (tylerflex) [reply tidy3d/components/simulation.py:186]: Also, add what happens if it's None?

File: tidy3d/components/data/sim_data.py [hunk 4]
```diff
@@ -145,7 +121,44 @@ def source_spectrum_fn(freqs):
...
+        if normalize_index and (normalize_index < 0 or normalize_index >= num_sources):
+            # normalize index out of bounds for source list
+            raise DataError(
+                f"normalize_index {normalize_index} out of bounds for list of sources "
+                f"of length {num_sources}"
+            )
+
+        def source_spectrum_fn(freqs):
+            """Normalization function that also removes previous normalization if needed."""
+            new_spectrum_fn = self.source_spectrum(normalize_index)
```
Comment (tylerflex) [inline tidy3d/components/data/sim_data.py:150]: if `normalize_index` or `self.simulation.normalize_index` is `None`, does this still work? or do we need to set `new_spectrum_fn` / `old_spectrum_fn` to `lambda x: 1.0` here? I'm not understanding this edge case.

**Review by tylerflex — CHANGES_REQUESTED** [review]
looks good overall, only thing I'm a bit unsure on are:
* Some of the validation logic (and None handling) is a little hard to follow, so let's just make sure we're testing all the edge cases.  For example, it seemed the `symmetry` might have never been applied because of that tiny bug.
* Sometimes `sim_data[name]` returns a copy (if symmetry has an effect) and sometimes it doesn't.  I wonder whether this inconsistency could create some confusion for the user.  Example: user doesn't understand why their modifications of `sim_data[mode].amps` modify the actual data whereas modifying `sim_data[field].Ex` has no lasting effect. At the very least, we should be very clear when things are copied, maybe even throw a warning in that case?

File: tidy3d/components/data/sim_data.py (hunk 4 above)
Comment (momchil-flex) [reply tidy3d/components/data/sim_data.py:150]: It works because of this https://github.com/flexcompute/tidy3d/blob/b263febca19d6a3764dcfb7a36ebdffa7f7174c8/tidy3d/components/data/sim_data.py#L102-L103

File: tidy3d/components/simulation.py (hunk 3 above)
Comment (momchil-flex) [reply tidy3d/components/simulation.py:186]: > * Some of the validation logic (and None handling) is a little hard to follow, so let's just make sure we're testing all the edge cases.  For example, it seemed the `symmetry` might have never been applied because of that tiny bug.

Right, I'm pretty sure the symmetry would have been caught by the backend tests (I'm yet to make this work with the backend), but I'll add more cases to the frontend tests too.

> * Sometimes `sim_data[name]` returns a copy (if symmetry has an effect) and sometimes it doesn't.  I wonder whether this inconsistency could create some confusion for the user.  Example: user doesn't understand why their modifications of `sim_data[mode].amps` modify the actual data whereas modifying `sim_data[field].Ex` has no lasting effect. At the very least, we should be very clear when things are copied, maybe even throw a warning in that case?

Yeah I was wondering about that too, but it seems redundant to make a copy without symmetry (and with symmetry, the original data is at least twice smaller, so it's not so bad). For your example, I guess we generally don't intend people to manually change the data, but I do agree it's not great to have the inconsistency. So you're suggesting raising a warning in the symmetry case when a copy is made? Or how else can we be very clear ( just in the docstrings?)

File: tidy3d/components/data/sim_data.py (hunk 4 above)
Comment (tylerflex) [reply tidy3d/components/data/sim_data.py:144]: ah ok, I see

File: tidy3d/components/simulation.py (hunk 3 above)
Comment (tylerflex) [reply tidy3d/components/simulation.py]: I'm not really sure, I think warning and docstrings is a good start, but may not be enough? is it problematic to apply symmetry when we load from file? and then the `SimulationData` in memory has symmetry (and normalization) both applied already and we won't have to worry about copying at all. Just a thought. 

Otherwise, our only other option (if we want to make things consistent) seems to be to copy everything, including when symmetry isn't applied.

1. Copy only when symmetry applied, warn and put note in docstrings.
2. Apply symmetry to everything when loading from file, copy nothing when accessing data.
3. Copy everything (like before).

**Comment by tylerflex** [comment]
Looks good besides the inconsistency issue still being discussed. Very minor comment but since `normalize_index` can be `None` sometimes, maybe we want the type annotation to be `normalize_index : Optional[int]`? But not sure if that's confusing since it's actually required. Maybe `normalize_index: Optional[int] = None` or `normalize_index: Optional[int] = Union[int, None]`?
